cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+cudf::type_dispatcher(data_type{type_id::FLOAT64}, size_of_functor{});  // returns 8
+```
+
+By default, `type_dispatcher` uses `cudf::type_to_id<t>` to provide the mapping of `cudf::type_id` 
+to dispatched C++ types. However, this mapping may be customized by explicitly specifying a 
+user-defined trait for the `IdTypeMap`. For example, to always dispatch `int32_t` for all values of 
+`cudf::type_id`:
+
+```c++
+template<cudf::type_id t> struct always_int{ using type = int32_t; }
+
+// This will always invoke `operator()<int32_t>`
+cudf::type_dispatcher<always_int>(data_type, f);
+```
+
+## Specializing Type-Dispatched Code Paths
+
+It is often necessary to customize the dispatched `operator()` for different types. This can be 
+done in several ways.
+
+The first method is to use explicit, full template specialization. This is useful for specializing 
+behavior for single types. The following example function object prints `"int32_t"` or `"double"` 
+when invoked with either of those types, or `"unhandled type"` otherwise.
+
+```c++
+struct type_printer {
+template <typename ColumnType>
+void operator()() { std::cout << "unhandled type\n"; }
+};
+
+// Due to a bug in g++, explicit member function specializations need to be
+// defined outside of the class definition
+template <>
+void type_printer::operator()<int32_t>() { std::cout << "int32_t\n"; }
+
+template <>
+void type_printer::operator()<double>() { std::cout << "double\n"; }
+```
+
+The second method is to use [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) with 
+`std::enable_if_t`. This is useful to partially specialize for a set of types with a common trait. 
+The following example functor prints `integral` or `floating point` for integral or floating point
+types, respectively.
+
+```c++
+struct integral_or_floating_point {
+template <typename ColumnType,
+          std::enable_if_t<not std::is_integral<ColumnType>::value and
+                           not std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "neither integral nor floating point\n"; }
+
+template <typename ColumnType,
+          std::enable_if_t<std::is_integral<ColumnType>::value>* = nullptr>
+void operator()() { std::cout << "integral\n"; }
+
+template < typename ColumnType,
+           std::enable_if_t<std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "floating point\n"; }
+};
+```
+
+For more info on SFINAE with `std::enable_if`, [see this post](https://eli.thegreenplace.net/2014/sfinae-and-enable_if).
+
+There are a number of traits defined in `include/cudf/utilities/traits.hpp` that are useful for 
+partial specialization of dispatched function objects. For example `is_numeric<T>()` can be used to 
+specialize for any numeric type.
+
+# Nested Data Types
+
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
```

Comment:
I wonder if we are conflating "variable size" types with "nested types" here:
1. Are strings really nested types?
2. `struct` is only of variable size if it (eventually) contains a string or list.

Reply:
>     1. Are strings really nested types?

This is a really slippery question.  To me, they are in limbo between nested and non-nested.  

The string column itself does indeed have child columns "nested" inside but that's more of an implementation detail than a data type. It's still logically just "strings".   On the other hand, 1 level list columns are exactly the same and those are clearly nested types.  

On the other, _other_ hand, I've found that when writing tests, I tend to only use the term "Nested" in the name when the test in question has >= two logical columns nested inside each other (ie, List<List<int>>).  If I'm just making a test with a basic list or struct, I tend to use "List" or "Struct" in the name.  

Arrow probably has something to say about this.

Reply:
I rephrased to "Variable-size and Nested Data Types".